### PR TITLE
Bf 334 support for selenium waits

### PR DESCRIPTION
**Pull request by beanqueen**

### CODE REVIEW

File: salad/steps/browser/waits.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
+# -*- coding: utf-8 -*-
+
+import re
+
+from lettuce import step, world
+
```
Comment (netme) [inline salad/steps/browser/waits.py]: No need to empty line here. Both libraries `lettuce` and `selenium` are 3rd party.

File: salad/steps/browser/waits.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    'with(?: the)? xpath "([^"]*)"': By.XPATH,
+    'with(?: the)? link text "([^"]*)"': By.LINK_TEXT,
+    'with(?: the)? partial link text "([^"]*)"': By.PARTIAL_LINK_TEXT,
+}
+THING_STRING = (
+    "(?:element|thing|field|textarea|radio button|button|checkbox|label|link)")
+
+
+def _construct_name(find_by, condition_string):
+        name = "wait_for_element_by_%s_to_%s" % (find_by, condition_string)
```
Comment (netme) [inline salad/steps/browser/waits.py]: better use `.format()` it's more readable.

File: salad/steps/browser/waits.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+               (THING_STRING, finder_string, condition_string))
+
+    @step(pattern)
+    def _this_step(step, selector, wait_time):
+        wait_time = int(wait_time or 10)
+        wait = WebDriverWait(world.browser.driver, wait_time)
+        try:
+            wait.until(expected_condition((find_by, selector),
+                                          expected_selection_state))
+        except Exception as e:
```
Comment (netme) [inline salad/steps/browser/waits.py]: Don't catch `Exception`. Please check which kind of exceptions `wait.until` hits and catch only them.

File: salad/steps/browser/waits.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+            _wait_for_selected_generator(finder_string, find_by,
+                                         condition_string, expected_condition,
+                                         expected_selection_state))
+
+
+# WAIT FOR ELEMENT TO HAVE A CERTAIN CONDITION
+def _wait_for_generator(finder_string, find_by,
+                        condition_string, expected_condition):
+    pattern = (r'wait for the %s %s to %s(?: within (\d+) seconds)?$' %
+               (THING_STRING, finder_string, condition_string))
```
Comment (netme) [inline salad/steps/browser/waits.py]: `.format()`

File: salad/steps/browser/waits.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+    @step(pattern)
+    def _this_step(step, selector, wait_time):
+        wait_time = int(wait_time or 10)
+        wait = WebDriverWait(world.browser.driver, wait_time)
+        try:
+            wait.until(expected_condition((find_by, selector)))
+        except Exception as e:
```
Comment (netme) [inline salad/steps/browser/waits.py]: The same with `Exception`

File: salad/steps/browser/waits.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    return _this_step
+
+
+# build the steps that wait for an element to have a certain condition
+for condition_string, expected_condition in WAIT_OPTIONS.iteritems():
+    for finder_string, find_by in ELEMENT_FINDERS.iteritems():
+        _wait_for_generator(finder_string, find_by,
+                            condition_string, expected_condition)
+        name = _construct_name(find_by, condition_string)
+        globals()[name] = (_wait_for_generator(finder_string, find_by,
```
Comment (netme) [inline salad/steps/browser/waits.py]: Not a nice approach. Better not to use global variables in this case.

File: salad/steps/browser/waits.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    pattern = (r'wait for the %s %s to %s(?: within (\d+) seconds)?$' %
+               (THING_STRING, finder_string, condition_string))
+
+    @step(pattern)
+    def _this_step(step, selector, text, wait_time):
+        wait_time = int(wait_time or 10)
+        wait = WebDriverWait(world.browser.driver, wait_time)
+        try:
+            wait.until(expected_condition((find_by, selector), text))
+        except Exception as e:
```
Comment (netme) [inline salad/steps/browser/waits.py]: The same with Exception

File: salad/steps/browser/waits.py (hunk 3 above)
Comment (emulbreh) [inline salad/steps/browser/waits.py]: yes

File: salad/steps/browser/waits.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+                   (finder_string, selector, condition_string, wait_time, e))
+            raise AssertionError(msg)
+    return _this_step
+
+
+# build the steps that wait for an element to have a certain selection state
+for condition_string, ec_tuple in WAIT_SELECTED_OPTIONS.iteritems():
+    for finder_string, find_by in ELEMENT_FINDERS.iteritems():
+        expected_condition = ec_tuple[0]
+        expected_selection_state = ec_tuple[1]
```
Comment (emulbreh) [inline salad/steps/browser/waits.py]: you can use unpacking here: ``expected_condition, expected_selection_state = ec_tuple``

File: salad/steps/browser/waits.py (hunk 5 above)
Comment (emulbreh) [inline salad/steps/browser/waits.py]: see above

File: salad/steps/browser/waits.py (hunk 7 above)
Comment (emulbreh) [inline salad/steps/browser/waits.py]: see above

**Comment by netme** [comment]
+1 :rocket: ship it

**Pull request merged**